tfhe/src/integer/server_key/radix_parallel/scalar_add.rs

```diff
@@ -255,7 +257,136 @@ impl ServerKey {
+
+                        let low_scalar =
+                            u64::from(packed_scalar_block) % self.message_modulus().0 as u64;
+                        let high_scalar =
+                            u64::from(packed_scalar_block) / self.message_modulus().0 as u64;
+
+                        let low_r = low_x + low_scalar;
+                        // The result of high_x + high_scalar will be shifted by one bit
+                        // so the bit at index 0 can accept the bit that encodes the propagation
+                        // state of the low block, such that when we add the carry generated
+                        // by the previous packed chunk we get the correct result of the high part
+                        //
+                        // That result is shifted by one bit, so the extract_message will need
+                        // to be different
+                        let v = if low_r >= self.message_modulus().0 as u64 {
+                            2
+                        } else if low_r == self.message_modulus().0 as u64 - 1 {
+                            1
+                        } else {
+                            0
+                        };
+
+                        (((high_scalar + high_x) % self.message_modulus().0 as u64) << 1) + v
+                    })
+                })
+                .collect::<Vec<_>>();
+
+            let mut inner_pack_propagation = packed_blocks.clone();
+            packed_blocks
+                .par_iter_mut()
+                .chain(inner_pack_propagation.par_iter_mut())
+                .zip(gpn_luts.par_iter().chain(luts2.par_iter()))
+                .for_each(|(block, lut)| self.key.apply_lookup_table_assign(block, lut));
+            let generates_or_propagates = packed_blocks; // Rename
+
+            let (carries, _) =
+                self.compute_carry_propagation_parallelized_low_latency(generates_or_propagates);
+
+            ct.blocks_mut()
+                .iter_mut()
+                .step_by(2)
+                .zip(
+                    BlockDecomposer::new(scalar, self.message_modulus().0.ilog2())
+                        .iter_as::<u8>()
+                        .step_by(2),
+                )
+                .for_each(|(block, scalar)| self.key.unchecked_scalar_add_assign(block, scalar));
+
+            ct.blocks_mut()
+                .iter_mut()
+                .skip(1)
+                .step_by(2)
+                .zip(inner_pack_propagation)
+                .for_each(|(ct_block, new_value)| *ct_block = new_value);
+
+            // Finally add the carries and then, do the message extract
+            ct.blocks_mut()
+                .chunks_mut(2)
+                .zip(carries.iter())
+                .for_each(|(blocks, carry)| {
+                    self.key.unchecked_add_assign(&mut blocks[0], carry);
+                    self.key.unchecked_add_assign(&mut blocks[1], carry);
+                });
```

Comment:
this confuses me 😵‍💫 let's maybe discuss quickly when you have some time so that I understand the algorithm better ?